sys/linux/bpf.txt

```diff
@@ -250,13 +251,21 @@ define BPF_JLE0	BPF_JLE >> 4
 define BPF_JSLT0	BPF_JSLT >> 4
 define BPF_JSLE0	BPF_JSLE >> 4
 
-bpf_insn_call {
+bpf_insn_call_helper {
 	code	const[bpf_call_code, int8]
 	regs	const[0, int8]
 	off	const[0, int16]
 	func	int32[0:__BPF_FUNC_MAX_ID]
 }
 
+bpf_insn_call_func {
+	code	const[bpf_call_code, int8]
+	dst	const[0, int8:4]
+	src	const[BPF_PSEUDO_CALL, int8:4]
+	off	const[0, int16]
+	func	int32
```

Comment:
Yes, the current language is not powerful enough to express this.
But perhaps it's better to do something like int32[0:8]. syzkaller should generally be generating relatively short programs, and referencing first 9 instructions should be able to cover all possible situations in the end. syzkaller will also treat 0:8 as a soft limit so eventually it will also pass -1 and other nonsense too.
Also please add a `# NEED:` comment describing what capability we need here for proper description. If  you do `grep NEED: sys/linux/*.txt` you will see all current cases, and it helps when we extend the language.

Reply:
> But perhaps it's better to do something like int32[0:8]. syzkaller should generally be generating relatively short programs, and referencing first 9 instructions should be able to cover all possible situations in the end. syzkaller will also treat 0:8 as a soft limit so eventually it will also pass -1 and other nonsense too.

Maybe `int32[-8:8]` in that case? The functions (called subprogs) can be before the call and having negative jump offsets could help fuzz the loop-detection part of the code.

A value of 0 is also not valid because calls are supposed to be to the start of a new subprog and the call instruction is always at least followed by a return instruction. But I don't think syzkaller has a syntax to express `[-8:-1] U [1:8]`? (And it would probably still try 0 as a special value anyway.)

> Also please add a `# NEED:` comment describing what capability we need here for proper description. If you do `grep NEED: sys/linux/*.txt` you will see all current cases, and it helps when we extend the language.

Will do.

Reply:
If this is an offset from the current insn, then int32[-8:8] makes sense.

Reply:
Yes, it is. I pushed an update branch.